kernel/src/process/signal/poll.rs

```diff
@@ -14,17 +17,39 @@ use crate::{
 
 /// A pollee that manages a set of [`IoEvents`].
 ///
-/// To correctly use the pollee, you need to call [`Pollee::notify`] whenever a new event arrives.
+/// To use the pollee correctly, you must follow the rules below carefully:
+///  * [`Pollee::notify`] needs to be called whenever a new event arrives.
+///  * [`Pollee::invalidate`] needs to be called whenever an old event disappears and no new event
+///    arrives.
```

Comment:
We can invalidate only when the buffer becomes empty, or we can invalidate every time we successfully read some data, or we can invalidate every time the user program makes a `recv` system call.

All are fine because the invalidation is allowed to be spurious.

Currently, TCP/UDP sockets and pipes will invalidate the pollee if some data are successfully read.

Reply:
> We can invalidate only when the buffer becomes empty, or we can invalidate every time we successfully read some data, or we can invalidate every time the user program makes a recv system call.

I think all code should follow the same pattern to make the code easy to check, alghouth they have the freedom to call this method at any time they want. 

> Currently, TCP/UDP sockets and pipes will invalidate the pollee if some data are successfully read.
 
This is reasonable. I hope all code using pollee with the same pattern, alghouth they are not to be included in this PR.

Reply:
Yeah, I agree with that. I was originally trying to make the usage consistent, but it's just tricky.

We should first agree on _which_ of the three mentioned in the comment is the desired pattern, and then refactor the corresponding code accordingly.

Given the facts that
 - before this PR, the way we update I/O events is also inconsistent (e.g. for pipes/sockets, they are updated after releasing the original locks, but for epoll file, it is updated while holding the ready list),
 - the new API is still compatible either way,

I just transfer the old code without bothering to refactor unrelated code, and we can leave other cleanup tasks for future work.